posts/2024-05-01-enabling-rust-lld-on-linux.md

```diff
@@ -0,0 +1,93 @@
+---
+layout: post
+title: Faster linking times on nightly on linux with `rust-lld`
+author: Rémy Rakic
+team: the compiler performance working group <https://www.rust-lang.org/governance/teams/compiler#team-wg-compiler-performance>
+---
+
+TL;DR: rustc will use `rust-lld` by default on `x86_64-unknown-linux-gnu` on nightly to
+significantly reduce linking times.
+
+#### Some context
+
+Linking time is often a big part of compilation time. When rustc needs to build a binary or a shared
+library, it will usually call the default linker installed on the system to do that (this can be
+changed on the command-line or by the target for which the code is compiled).
+
+The linkers do an important job, with concerns about stability, backwards-compatibility and so on.
+For these and other reasons, on the most popular operating systems they usually are older programs,
+designed when computers only had a single core. So, they usually tend to be slow on a modern
+machine. For example, when building ripgrep 13 in debug mode on linux, roughly half of the time is
+actually spent in the linker.
+
+There are different linkers, however, and the usual advice to improve linking times is to use one of
+these newer and faster linkers, like LLVM's [`lld`](https://lld.llvm.org/) or Rui Ueyama's
+[`mold`](https://github.com/rui314/mold).
+
+Some of rust's wasm and aarch64 targets already use `lld` by default. When using rustup, rustc ships
+with a version of `lld` for this purpose. When CI builds LLVM to use in the compiler, it also builds
+the linker and packages it. It's referred to as `rust-lld` to avoid colliding with any `lld` already
+installed on the user's machine.
+
+Since improvements to linking times are substantial, it would be a good default to use in the most
+popular targets. This has been discussed for a long time, for example in issues
+[#39915](https://github.com/rust-lang/rust/issues/39915) and
+[#71515](https://github.com/rust-lang/rust/issues/71515), and rustc already offers nightly flags to
+use `rust-lld`.
+
+By now, we believe we've done all the internal testing that we could, on CI, crater, our
+benchmarking infrastructure, and would like to expand testing and gather real-world feedback and
+use-cases. Therefore, we will enable `rust-lld` to be the linker used by default on
```

Comment:
```suggestion
By now, we believe we've done all the internal testing that we could, on CI, crater, and our
benchmarking infrastructure. We would now like to expand testing and gather real-world feedback and
use-cases. Therefore, we will enable `rust-lld` to be the linker used by default on
```

(The text as originally written had a list of three items and then sort of fell off into the "next step" description without a clear signal that this wasn't part of the three item list. This suggestion is an attempt to fix that.)

Reply:
Thanks